File: backend/src/main/java/com/pawportal/backend/services/implementations/DogService.java
package com.pawportal.backend.services.implementations;

import com.pawportal.backend.models.DogModel;
import com.pawportal.backend.models.enums.DogStatus;
import com.pawportal.backend.repositories.DogRepository;
import com.pawportal.backend.services.interfaces.IDogService;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/** Developed by Group 6:
 * Kenji Mark Alan Arceo
 * Carl Norbi Felonia
 * Ryonan Owen Ferrer
 * Dino Alfred Timbol
 * Mike Emil Vocal
 */

/**
 * DogService class. Deals with dog related service operations
 */

@Service
@RequiredArgsConstructor
public class DogService implements IDogService {

    private final DogRepository dogRepository;

    //Retrieves all dogs
    @Override
    public List<DogModel> getAllDogs() {
        return dogRepository.findAll();
    }

    //Retrieves a particular dog
    @Override
    public Optional<DogModel> getDogById(Long id) {
        return dogRepository.findById(id);
    }

    //Retrieves dogs according to status (used for filtering)
    @Override
    public List<DogModel> getDogsByStatus(DogStatus status) {
        return dogRepository.findByStatus(status);
    }

    //Creates a dog
    @Override
    public DogModel createDog(DogModel dog) {
        return dogRepository.save(dog);
    }

    //Updates dog information
    @Override
    public DogModel updateDog(Long id, DogModel dogDetails) {
        DogModel dog = dogRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Dog not found with id: " + id));

        dog.setName(dogDetails.getName());
        dog.setDescription(dogDetails.getDescription());
        dog.setBreed(dogDetails.getBreed());
        dog.setAge(dogDetails.getAge());
        dog.setSize(dogDetails.getSize());
        dog.setGender(dogDetails.getGender());
        dog.setImageUrl(dogDetails.getImageUrl());
        dog.setTemperament(dogDetails.getTemperament());
        dog.setVaccinated(dogDetails.getVaccinated());
        dog.setSpayedNeutered(dogDetails.getSpayedNeutered());
        dog.setStatus(dogDetails.getStatus());

        return dogRepository.save(dog);
    }

    //Deletes dog information
    @Override
    public void deleteDog(Long id) {
        dogRepository.deleteById(id);
    }
}

File: frontend/src/app/services/dog.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Dog } from '../models/dog.model';
import { TokenStorageService } from './token-storage.service';

/**
 * Developed by Group 6:
    Kenji Mark Alan Arceo
    Carl Norbi Felonia
    Ryonan Owen Ferrer
    Dino Alfred Timbol
    Mike Emil Vocal
 */

/**
 * Service class for dog operations
 */
@Injectable({
  providedIn: 'root'
})
export class DogService {
  private apiUrl = 'http://localhost:8080/dogs';

  constructor(
    private http: HttpClient,
    private tokenStorageService: TokenStorageService
  ) {}

  private getHeaders(): HttpHeaders {
    const token = this.tokenStorageService.getToken();
    return new HttpHeaders({
      'Authorization': `Bearer ${token}`,
      'Content-Type': 'application/json'
    });
  }

  getAllDogs(): Observable<Dog[]> {
    return this.http.get<Dog[]>(this.apiUrl,
    );
  }

  getDogById(id: number): Observable<Dog> {
    return this.http.get<Dog>(`${this.apiUrl}/${id}`,
      { headers: this.getHeaders() }
    );
  }

  getAvailableDogs(): Observable<Dog[]> {
    return this.http.get<Dog[]>(`${this.apiUrl}/status/AVAILABLE`,
    );
  }

  createDog(dog: Dog): Observable<Dog> {
    return this.http.post<Dog>(this.apiUrl, dog, 
      { headers: this.getHeaders() });
  }

  updateDog(id: number, dog: Dog): Observable<Dog> {
    return this.http.put<Dog>(`${this.apiUrl}/${id}`, dog, 
      { headers: this.getHeaders() });
  }

  deleteDog(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`, 
      { headers: this.getHeaders() });
  }
}
